src/container/container.ts

```diff
@@ -593,15 +592,15 @@ class Container implements interfaces.Container {
     private _getNotAllArgs<T>(
         serviceIdentifier: interfaces.ServiceIdentifier<T>,
         isMultiInject: boolean,
-        key?: string | number | symbol,
-        value?: any,
+        tags?: interfaces.Tag[]
     ): GetArgs<T> {
         const getNotAllArgs: GetArgs<T> = {
             avoidConstraints: false,
             isMultiInject,
             serviceIdentifier,
-            key,
-            value,
+            key: tags?.[0]?.[0],
+            value: tags?.[0]?.[1],
+            tags: tags?.slice(1)
```

Comment:
This is a problem.

```

        it("Should provide all tags in a middleware", () => {
            const middleware:interfaces.Middleware = (next => {
                return (nextArgs => {
                    if(nextArgs.tags!.length !== 2){
                        throw new Error("Where's the tag !?")
                    }
                    return next(nextArgs);
                })
            })

            const container = new Container();
            container.applyMiddleware(middleware);
            expect(container.getTagged<string>("Sid", [["tag1", 1],["tag2",2]])).to.equal("Two");
        })
```

Reply:
Do you have any suggestion on how to solve this without breaking the other test you provided?

I've tried the solution with getters and setters but it was a tricky one to solve. I'll invest more time on this.

Reply:
May have missed something myself.  First attempt

```

    private _getNotAllArgs<T>(
        serviceIdentifier: interfaces.ServiceIdentifier<T>,
        isMultiInject: boolean,
        tags?: interfaces.Tag[]
    ): GetArgs<T> {
        const getNotAllArgs: GetArgs<T> = {
            avoidConstraints: false,
            isMultiInject,
            serviceIdentifier,
            key: tags?.[0]?.[0],
            value: tags?.[0]?.[1],
            tags: tags??[]
        };

        return getNotAllArgs;
    }



    private _planAndResolve<T>(): (args: interfaces.NextArgs<T>) => interfaces.ContainerResolution<T> {
        return (args: interfaces.NextArgs<T>) => {
            if (args.tags.length> 0 ){
                const tags = [...args.tags];
                const firstTag = tags.shift()
                if(args.key !== firstTag![0] || args.value !== firstTag![1]){
                    if (args.key !== undefined) {
                        args.tags = [[args.key,args.value],...tags]
                    }else {
                        args.tags = tags;
                    }
                }
            }
```

Reply:
The problem is that if change the array of `tags` and do not change the `key` and `value` this code will override the values with the original value.

Reply:
Indeed

Reply:
could store the original args on the container for checking in planAndResolve if that helps

Reply:
Thinking aloud

```
private _get<T>(getArgs: GetArgs<T>): interfaces.ContainerResolution<T> {
        this.tags = [...getArgs.tags];


    private tagsChanged(tags:interfaces.Tag[]):boolean{
        let tagsDidChange = false;
        if(this.tags.length === tags.length){
            for(let i =0;i<this.tags.length;i++){
                if(this.tags[i]!== tags[i]){
                    tagsDidChange = true;
                    break;
                }
            }
        }else{
            tagsDidChange = true;
        }
        return tagsDidChange;
    }


    private _planAndResolve<T>(): (args: interfaces.NextArgs<T>) => interfaces.ContainerResolution<T> {
        return (args: interfaces.NextArgs<T>) => {
            if (!this.tagsChanged(args.tags)){
                const tags = [...args.tags];
                const firstTag = tags.shift()
                if(args.key !== firstTag![0] || args.value !== firstTag![1]){
                    if (args.key !== undefined) {
                        args.tags = [[args.key,args.value],...tags]
                    }else {
                        args.tags = tags;
                    }
                }
            }

```

Reply:
I'll experiment with that idea.

Reply:
I think was able to figure it out using getters and setters.

Reply:
The code needs to work with the NextArgs interface.  Getters and setters do not work if they are not used.  The middleware can return a different object.

I think storing on the Container is the only solution available.  
A corrected version of my idea from before.

```
private _get<T>(getArgs: GetArgs<T>): interfaces.ContainerResolution<T> {
        this.tags = getArgs.tags.map(tag => [tag[0],tag[1]]);


    private _planAndResolve<T>(): (args: interfaces.NextArgs<T>) => interfaces.ContainerResolution<T> {
        return (args: interfaces.NextArgs<T>) => {
            this._mapTagsFromKeyValue(args);

    private _tagsChanged(tags:interfaces.Tag[]):boolean{
        return this.tags.length !== tags.length ? true : this.tags.some((originalTag,i) => {
            const newTag = tags[i];
            return originalTag[0] !== newTag[0] || originalTag[1] !== newTag[1];
        });
    }
    
    private _mapTagsFromKeyValue<T>(args: interfaces.NextArgs<T>):void {
        if (!this._tagsChanged(args.tags)){
            if(this.tags.length > 0) {
                const tags = [...args.tags];
                const firstTag = tags.shift() as interfaces.Tag;
                if(args.key !== firstTag[0] || args.value !== firstTag[1]){
                    args.tags = [[args.key!,args.value],...tags];
                }
            } else {
                if(args.key !== undefined) {
                    args.tags = [[args.key,args.value]];
                }
            }
        }
    }
```

tests used
```

    it("should provide key and value from first tag for old middleware", () => {
        let key:interfaces.NextArgs<unknown>["key"];
        let value:interfaces.NextArgs<unknown>["value"];
        const middleware:interfaces.Middleware = (next => {
            return (nextArgs => {
                key = nextArgs.key;
                value = nextArgs.value;
                return next(nextArgs);
            })
        })

        const container = new Container();
        container.bind("Sid").toConstantValue("Value").whenTargetMultiTagged(["tag1", 1],["tag2",2]);
        container.applyMiddleware(middleware);
        container.getTagged<string>("Sid", [["tag1", 1],["tag2",2]]);
        expect(key).to.equal("tag1");
        expect(value).to.equal(1);
    });

    it("should work with old middleware that changes value", () => {
        const middleware:interfaces.Middleware = (next => {
            return (nextArgs => {
                if(nextArgs.key === "tag" && nextArgs.value === 1){
                    nextArgs.value = 2;
                }
                return next(nextArgs);
            })
        })

        const container = new Container();
        container.applyMiddleware(middleware);
        container.bind('Sid').toConstantValue("One").whenTargetTagged("tag",1);
        container.bind('Sid').toConstantValue("Two").whenTargetTagged("tag",2);
        expect(container.getTagged<string>("Sid","tag",1)).to.equal("Two");
    });

    it("should work with old middleware that changes key", () => {
        const middleware:interfaces.Middleware = (next => {
            return (nextArgs => {
                nextArgs.key = "new_key";
                return next(nextArgs);
            })
        })

        const container = new Container();
        container.applyMiddleware(middleware);
        container.bind('Sid').toConstantValue("One").whenTargetTagged("tag",1);
        container.bind('Sid').toConstantValue("Two").whenTargetTagged("new_key",1);
        expect(container.getTagged<string>("Sid","tag",1)).to.equal("Two");
    });

    it("should work with old middleware that adds a key", () => {
        const middleware:interfaces.Middleware = (next => {
            return (nextArgs => {
                nextArgs.key = "tag";
                nextArgs.value = 2;
                return next(nextArgs);
            })
        })

        const container = new Container();
        container.applyMiddleware(middleware);
        container.bind('Sid').toConstantValue("One").whenTargetTagged("tag",1);
        container.bind('Sid').toConstantValue("Two").whenTargetTagged("tag",2);
        expect(container.get<string>("Sid")).to.equal("Two");
    });

    it("Should provide all tags for new middleware", () => {
        let tags:interfaces.Tag[]|undefined
        const middleware:interfaces.Middleware = (next => {
            return (nextArgs => {
                tags = nextArgs.tags;
                return next(nextArgs);
            })
        })

        const container = new Container();
        container.bind("Sid").toConstantValue("Value").whenTargetMultiTagged(["tag1", 1],["tag2",2]);
        container.applyMiddleware(middleware);
        container.getTagged<string>("Sid", [["tag1", 1],["tag2",2]]);
        expect(tags).to.eql([["tag1", 1],["tag2",2]]);
    });

    it("Should work with new middleware that changes key", () => {
        const middleware:interfaces.Middleware = (next => {
            return (nextArgs => {
                const firstTag = nextArgs.tags[0];
                firstTag[0] = "new_key";
                return next(nextArgs);
            })
        })

        const container = new Container();
        container.bind("Sid").toConstantValue("X").whenTargetMultiTagged(["new_key",1],["tag2",2]);
        container.applyMiddleware(middleware);
        expect(container.getTagged<string>("Sid", [["tag1", 1],["tag2",2]])).to.equal("X");
    });

    it("Should work with new middleware that changes value", () => {
        const middleware:interfaces.Middleware = (next => {
            return (nextArgs => {
                const firstTag = nextArgs.tags[0];
                firstTag[1] = "New value";
                return next(nextArgs);
            })
        })

        const container = new Container();
        container.bind("Sid").toConstantValue("X").whenTargetMultiTagged(["tag1","New value"],["tag2",2]);
        container.applyMiddleware(middleware);
        expect(container.getTagged<string>("Sid", [["tag1", 1],["tag2",2]])).to.equal("X");
    });

    it("Should work with new middleware that changes the size of tags array", () => {
        const middleware:interfaces.Middleware = (next => {
            return (nextArgs => {
                nextArgs.tags.push(["tag2",2]);
                return next(nextArgs);
            })
        })

        const container = new Container();
        container.bind("Sid").toConstantValue("X").whenTargetMultiTagged(["tag1",1],["tag2",2]);
        container.applyMiddleware(middleware);
        expect(container.getTagged<string>("Sid", [["tag1", 1]])).to.equal("X");
    });

    it("Should work with middleware returning own NextArgs", () => {
        const middleware:interfaces.Middleware = (next => {
            return (nextArgs => {
                nextArgs= {...nextArgs};
                if(nextArgs.key === "tag" && nextArgs.value === 1){
                    nextArgs.value = 2;
                }
                return next(nextArgs);
            })
        })

        const container = new Container();
        container.applyMiddleware(middleware);
        container.bind('Sid').toConstantValue("One").whenTargetTagged("tag",1);
        container.bind('Sid').toConstantValue("Two").whenTargetTagged("tag",2);
        expect(container.getTagged<string>("Sid","tag",1)).to.equal("Two");
    });

```

I suggest that you look for a flaw in that and if you cannot find one then use it.

Reply:
I though about using the container to store the tags, but wouldn't that be a problem when resolving things async?

You could have two resolutions happening at the same time and resetting the original tags before the other has finished.

I think that the probably the best way is not to memoize the set of middlewares and reuse it but create a chain of middlewares for each resolution. In that way, we can inject the original tags directly in `_planAndResolve`.

Reply:
> I though about using the container to store the tags, but wouldn't that be a problem when resolving things async?

Yes ignore that idea doesn't work for nested gets.

Can we use a stack of tags ?

Reply:
Middleware is not async.  We call a "get" method - _get sets up the tags and the args goes through the middleware until the last middleware _planAndResolve gets to fix up before resolution.  No other code can change the tags on the container before they are used.

Unless I am missing something

Reply:
But a middleware could return anything, right?

The last middleware is the `planAndResolve` that could return a `Promise`. Let's say a middleware is returning a promise directly instead of calling the next function. Or waiting for a promise then calling the next function. We could reuse the tags from other requests

```Typescript
const myMiddleware = (next) => async (nextArgs) => {
  await somePromise()
  return next(nextArgs)
}
```

This would work because of how we are checking the result of the middleware chain, right? The next in this middleware could be `planAndResolve` and it could read the tags from another request.

Reply:
> But a middleware could return anything, right?

No.

```
export type Middleware = (next: Next) => Next;
export type Next = (args: NextArgs) => (any | any[]);
```

Reply:
If the middleware decided to kick off another "get" then putting the tags on the container would no work

Reply:
Why not building the whole chain of middlewares for each request and inject the tags in the `planAndResolve`?

Reply:
I assume that you mean that applyMiddleware just stores the middleware and the _get method does what applyMiddleware did before but allows fixing the original args to be consumed by planAndResolve ?

Only problem is that this is a breaking change.

Reply:
Why is that a breaking change?

Reply:
Because you are invoking the middleware for every request instead of a one time setup.

Reply:
I see. It breaks the expectation from the user that it would be invoked only once.

Reply:
Perhaps then a middleware wrapping function with original nextargs and actual. Create these in applyMiddleware. Have the last in the chain no longer be planAndResolve but the fix up that will then call planAndResolve.

Reply:
That would be a breaking change as well, wouldn't it? If the middleware creator is resolving another value and returning it we would ignore it.

Reply:
Will post what I think works tonight or tomorrow morning

Reply:
I'll try to make it work with the idea of the stack.

Reply:
Does it already ? tagStack is just an array of tags.

Reply:
I've tried to implement using a version with plan and resolve outside the middleware chain. There is a problem with allowing middlewares to handle errors in resolution =(

> Doesn't it already ?! tagstack is just an array of tags.

tagStack solution is risky when a middleware code throws. We would need to be able to clean the stack. And I feel like holding state for the plan is not ideal.

Reply:
Will be an issue with how it determines if tags changed

Reply:
> feel like holding state for the plan is not ideal.

Certainly not ideal !

Reply:
As for the error we can just handle clear the stack and rethrow

Reply:
> As for the error we can just handle clear the stack and rethrow

I don't think that works, because at the point that we would throw the exception, the executing of the function that needs to be able to handle has already finished.

I have another idea, my last one. If that doesn't work, I'll close this pr and move this feature to `v6` and make it a breaking change removing `key` and `value` from the args.